[C10D] Fix pointToPoint op Flight Recording

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120430
* #120745
* __->__ #120270
* #120724

Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

Another approach also ditched
- record events on workEnqueue instead of initWork
- reason ditched: too messy to get input/output shapes tagged on
  recording when recording in workEnqueue.  Adding the info onto the
  Work obj would be possible, but adds to overhead of copying Works
  which we do on every collective. We can get info off the input/output
  tensors directly in initWork, but we don't want to keep refs to those
  tensors alive while the work is Enqueued, so we'd have to specifically
  copy size lists or something.

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes initWork to accept a 'record' argument. This defaults to
false, and should only be set to true if the caller ensures the work
will be enqueued by workEnqueue, ensuring its cuda events are live when
used by flight recorder's update_state().

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225

wconstab committed:
Try fixing pointToPoint to not enqueue work if it's coalescing

this approach also seems possible.  upside is we no longer create a
pointless work, and we can keep the initWork 1:1 with flight recorder record().

downside is this is a bigger change and a bit risky.  also not sure why
endCoalescing was only enqueueing works for CoalColl not CoalP2P

[ghstack-poisoned]

wconstab committed:
Update on "Try fixing pointToPoint to not enqueue work if it's coalescing"


this approach also seems possible.  upside is we no longer create a
pointless work, and we can keep the initWork 1:1 with flight recorder record().

downside is this is a bigger change and a bit risky.  also not sure why
endCoalescing was only enqueueing works for CoalColl not CoalP2P

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab committed:
Update on "Try fixing pointToPoint to not enqueue work if it's coalescing"


this approach also seems possible.  upside is we no longer create a
pointless work, and we can keep the initWork 1:1 with flight recorder record().

downside is this is a bigger change and a bit risky.  also not sure why
endCoalescing was only enqueueing works for CoalColl not CoalP2P

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab committed:
Update on "Try fixing pointToPoint to not enqueue work if it's coalescing"


this approach also seems possible.  upside is we no longer create a
pointless work, and we can keep the initWork 1:1 with flight recorder record().

downside is this is a bigger change and a bit risky.  also not sure why
endCoalescing was only enqueueing works for CoalColl not CoalP2P

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab committed:
Update on "Try fixing pointToPoint to not enqueue work if it's coalescing"


this approach also seems possible.  upside is we no longer create a
pointless work, and we can keep the initWork 1:1 with flight recorder record().

downside is this is a bigger change and a bit risky.  also not sure why
endCoalescing was only enqueueing works for CoalColl not CoalP2P

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab committed:
Update on "Try fixing pointToPoint to not enqueue work if it's coalescing"


this approach also seems possible.  upside is we no longer create a
pointless work, and we can keep the initWork 1:1 with flight recorder record().

downside is this is a bigger change and a bit risky.  also not sure why
endCoalescing was only enqueueing works for CoalColl not CoalP2P

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix pointToPoint op Flight Recording"


Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes the default flight recording to record only on
enqueueWork rather than inside initWork.  This should be a safer policy
in general as we assume cuda events captured during record() will be
kept alive by work, but that lifetime is only gauranteed if the work was
enqueued.

A downside is that we have to smuggle additional tensor shape
and profilingTitle along with the Work object if we want to record it.
- Open question- how to resolve this?

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix pointToPoint op Flight Recording"


Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes the default flight recording to record only on
enqueueWork rather than inside initWork.  This should be a safer policy
in general as we assume cuda events captured during record() will be
kept alive by work, but that lifetime is only gauranteed if the work was
enqueued.

A downside is that we have to smuggle additional tensor shape
and profilingTitle along with the Work object if we want to record it.
- Open question- how to resolve this?

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix pointToPoint op Flight Recording"


Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes the default flight recording to record only on
enqueueWork rather than inside initWork.  This should be a safer policy
in general as we assume cuda events captured during record() will be
kept alive by work, but that lifetime is only gauranteed if the work was
enqueued.

A downside is that we have to smuggle additional tensor shape
and profilingTitle along with the Work object if we want to record it.
- Open question- how to resolve this?

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix pointToPoint op Flight Recording"


Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

Another approach also ditched
- record events on workEnqueue instead of initWork
- reason ditched: too messy to get input/output shapes tagged on
  recording when recording in workEnqueue.  Adding the info onto the
  Work obj would be possible, but adds to overhead of copying Works
  which we do on every collective. We can get info off the input/output
  tensors directly in initWork, but we don't want to keep refs to those
  tensors alive while the work is Enqueued, so we'd have to specifically
  copy size lists or something.

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes initWork to accept a 'record' argument. This defaults to
false, and should only be set to true if the caller ensures the work
will be enqueued by workEnqueue, ensuring its cuda events are live when
used by flight recorder's update_state().

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

[ghstack-poisoned]

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2300,22 +2320,21 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::endCoalescing(OpType optype) {
+  // Bump collective counter after creating the work-
+  // this way all the flight-recorder events inside the coalesing group share
+  // the seq_ with the coalsed op and the work that goes with it
+  seq_++;
```
Should we bump the seq_ only if enqueue is true? my understanding is that there is a one to one mapping of 
seq_ -> enqueued work?

wconstab replied:
well, see my latest version and then lets talk. I wonder if you will agree with my new approach or not.  (didn't see your comment before i pushed.)

wconstab committed:
Update on "[C10D] Fix pointToPoint op Flight Recording"


Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

Another approach also ditched
- record events on workEnqueue instead of initWork
- reason ditched: too messy to get input/output shapes tagged on
  recording when recording in workEnqueue.  Adding the info onto the
  Work obj would be possible, but adds to overhead of copying Works
  which we do on every collective. We can get info off the input/output
  tensors directly in initWork, but we don't want to keep refs to those
  tensors alive while the work is Enqueued, so we'd have to specifically
  copy size lists or something.

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes initWork to accept a 'record' argument. This defaults to
false, and should only be set to true if the caller ensures the work
will be enqueued by workEnqueue, ensuring its cuda events are live when
used by flight recorder's update_state().

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix pointToPoint op Flight Recording"


Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

Another approach also ditched
- record events on workEnqueue instead of initWork
- reason ditched: too messy to get input/output shapes tagged on
  recording when recording in workEnqueue.  Adding the info onto the
  Work obj would be possible, but adds to overhead of copying Works
  which we do on every collective. We can get info off the input/output
  tensors directly in initWork, but we don't want to keep refs to those
  tensors alive while the work is Enqueued, so we'd have to specifically
  copy size lists or something.

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes initWork to accept a 'record' argument. This defaults to
false, and should only be set to true if the caller ensures the work
will be enqueued by workEnqueue, ensuring its cuda events are live when
used by flight recorder's update_state().

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

[ghstack-poisoned]

wconstab commented:
@zdevito one thing I forgot to do was add the extra ID to flight entries we discussed.

i think i can just do this inside the record function.  basically, an ID that increments on every record() call even if seq_ can be repeated.  Any issue with that approach?

XilunWu commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -4257,6 +4258,127 @@ def gather_trace():
+
+            if timing_enabled:
+                duration = t['entries'][seq]['duration_ms']
+                self.assertTrue(0.001 < duration < 10000, duration)
```
ditto

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2202,15 +2203,30 @@ c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
+    // may go
+    //   out of scope before flight recorder has retired them,
+    //   so we must ensure that any work that is initialized via initWork will
+    //   be enqueued
```
Do we guarantee "any work that is initialized via initWork will be enqueued'? Since I see a few calls of " initWork(device, rank_, optype, "nccl:coalesced", {}, {}, enqueue);", when  'enqueue' is false, work is inited but will not be enqueued

shuqiangzhang replied:
Maybe it is clearer if we can logically separate the concept of 'record' from 'work'.   we can call NCCLTraceBuffer::get()->record anytime, e.g, outside of initWork, when ever we want to add a flight record, and it might or might not be to a 'Work'

wconstab replied:
we can change the design to decouple record() from work, but that is a bigger scope than this PR.  This PR attempts to just make the current design more explicit.

The complexity comes from (1) record takes a ref to cuda events owned by work and lifecycle managed by work, (2) work lifecycle managed by enqueue and watchdog loop, (3) flight recorder 'retire' ends lifecycle after enqueued work is procesesed in watchdog, (4) i think @zdevito intentionally avoided making new cuda events for recorder vs work, for perf reasons

So, with the current design we should ensure any time a record happens, its work is kept alive.  But i did not add a 'mechanism' for this, just wrote out the rules.  Open to suggestions for how to do this.

shuqiangzhang replied:
Agree that we should not making new cuda events. But I saw sometimes record is already happen outside of initwork intentionally, we can make this style consistent?
<img width="732" alt="image" src="https://github.com/pytorch/pytorch/assets/4711970/86af2dff-f960-4a15-8aa6-04ee3ca079ea">

shuqiangzhang replied:
Then, I think it is fine to keep this way for now, I just found it a bit hard to wrap my head on which case we should record with initWork and which cases we should not

XilunWu reviewed (commented):
left some questions. I feel a bit difficult in understanding the change because it requires background knowledge in:
1) seq_num bumping in the original PG code
2) the work queue, which is necessary to understand the interaction of `record()` and work queue

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -4257,6 +4258,127 @@ def gather_trace():
+
+            if timing_enabled:
+                duration = t['entries'][coalesced_op]['duration_ms']
+                self.assertTrue(0.001 < duration < 10000, duration)
```
this doesn't come from anything. I just made it up.  It is just supposed to be 'sufficiently wide bounds that it will always pass if a real collective ran, but will catch a zero value that could happen if no collective ran, or a garbage value if we emitted some kind of corrupted data

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -4257,6 +4258,127 @@ def gather_trace():
+                if self.rank == 0:
+                    dist.recv(tensor, 1)
+                elif self.rank == 1:
+                    tensor *= 2
```
just for fun :) (well, actually, more so that there is a different value and we're sure we actually did something during the send)

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2264,6 +2280,15 @@ void ProcessGroupNCCL::startCoalescing() {
+  // inside endCoalescing, but before initWork. Since we now record individual
+  // ops from a coalesce group into the flight recorder, we want to have the
+  // same seq_ for those ops and its 'endCoalescing' op. Hence we bump during
+  // start, which has one minor downside- we burn a seq_ if someone ever does a
+  // 'start' and 'end' coalescing region without doing an operation inbetween.
```
yes, it is possible to start/end without doing a collective, so, an empty coalesce.

wconstab committed:
Update on "[C10D] Fix pointToPoint op Flight Recording"


Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

Another approach also ditched
- record events on workEnqueue instead of initWork
- reason ditched: too messy to get input/output shapes tagged on
  recording when recording in workEnqueue.  Adding the info onto the
  Work obj would be possible, but adds to overhead of copying Works
  which we do on every collective. We can get info off the input/output
  tensors directly in initWork, but we don't want to keep refs to those
  tensors alive while the work is Enqueued, so we'd have to specifically
  copy size lists or something.

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes initWork to accept a 'record' argument. This defaults to
false, and should only be set to true if the caller ensures the work
will be enqueued by workEnqueue, ensuring its cuda events are live when
used by flight recorder's update_state().

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix pointToPoint op Flight Recording"


Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

Another approach also ditched
- record events on workEnqueue instead of initWork
- reason ditched: too messy to get input/output shapes tagged on
  recording when recording in workEnqueue.  Adding the info onto the
  Work obj would be possible, but adds to overhead of copying Works
  which we do on every collective. We can get info off the input/output
  tensors directly in initWork, but we don't want to keep refs to those
  tensors alive while the work is Enqueued, so we'd have to specifically
  copy size lists or something.

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes initWork to accept a 'record' argument. This defaults to
false, and should only be set to true if the caller ensures the work
will be enqueued by workEnqueue, ensuring its cuda events are live when
used by flight recorder's update_state().

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2688,43 +2753,45 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
-    work->future_->markCompleted(at::IValue(*work->outputs_));
-  }
+    // End event should only be recorded after the ncclGroupEnd()
+    if (!coalescing_state_) {
```
Redundant check.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -319,8 +319,10 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+    // Store a reference to NCCL collective's outputs and inputs, used by
+    // result, to give a more descriptive message when representing the Work as
+    // a string, and for Flight Recorder.
+    std::shared_ptr<std::vector<at::Tensor>> inputs_;
```
I can see that FR needs `inputs_`, but it seems we can satisfy it by passing `inputs` to the `initWork` function (where `record()` occurs). But does that mean we'd also need to create an `inputs_` field in the `Work` object? I'd prefer having as few reference to tensor as possible. Or do I miss some other purpose?

wconstab replied:
oops that was leftover from one of my previous attempts.  i just removed it.

wconstab committed:
Update on "[C10D] Fix pointToPoint op Flight Recording"


Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

Another approach also ditched
- record events on workEnqueue instead of initWork
- reason ditched: too messy to get input/output shapes tagged on
  recording when recording in workEnqueue.  Adding the info onto the
  Work obj would be possible, but adds to overhead of copying Works
  which we do on every collective. We can get info off the input/output
  tensors directly in initWork, but we don't want to keep refs to those
  tensors alive while the work is Enqueued, so we'd have to specifically
  copy size lists or something.

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes initWork to accept a 'record' argument. This defaults to
false, and should only be set to true if the caller ensures the work
will be enqueued by workEnqueue, ensuring its cuda events are live when
used by flight recorder's update_state().

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix pointToPoint op Flight Recording"


Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

Another approach also ditched
- record events on workEnqueue instead of initWork
- reason ditched: too messy to get input/output shapes tagged on
  recording when recording in workEnqueue.  Adding the info onto the
  Work obj would be possible, but adds to overhead of copying Works
  which we do on every collective. We can get info off the input/output
  tensors directly in initWork, but we don't want to keep refs to those
  tensors alive while the work is Enqueued, so we'd have to specifically
  copy size lists or something.

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes initWork to accept a 'record' argument. This defaults to
false, and should only be set to true if the caller ensures the work
will be enqueued by workEnqueue, ensuring its cuda events are live when
used by flight recorder's update_state().

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix pointToPoint op Flight Recording"


Fix and test issues with both coalesced and individual send/recv ops

Considered an alternate approach and then ditched it
 - alternate approach: #119757
 - reason ditched: prefer recording individual collective events inside
   coalescing region instead of just the event at the end of the region,
   which also would not have tensor sizes or opnames without additional
   state variables added

Another approach also ditched
- record events on workEnqueue instead of initWork
- reason ditched: too messy to get input/output shapes tagged on
  recording when recording in workEnqueue.  Adding the info onto the
  Work obj would be possible, but adds to overhead of copying Works
  which we do on every collective. We can get info off the input/output
  tensors directly in initWork, but we don't want to keep refs to those
  tensors alive while the work is Enqueued, so we'd have to specifically
  copy size lists or something.

This PR instead avoids creating a work inside pointToPoint when
coalescing is active. Instead, only at endCoalescing() is a work finally
intialized and enqueued.  But it adds a record() call inside
pointToPoint() instead of creating a work, during coalescing. This
record() call picks up tensor shapes and op names.

It ALSO changes initWork to accept a 'record' argument. This defaults to
false, and should only be set to true if the caller ensures the work
will be enqueued by workEnqueue, ensuring its cuda events are live when
used by flight recorder's update_state().

The testing uncovers some odd pre-existing behavior and leaves them
alone for now. We could change some of these
- seq starts off at 1, not 0 for first op (but this is inconistent)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]
